Feat: Deployment action templates

chrismclarke opened the pull request:
PR Checklist

- [x] PR title descriptive (can be used in release notes)

## Description

Adds templating system to make it easier to configure github actions for individual deployments

## Review Notes
- Set a deployment to configure github actions for
- Run the command `yarn workflow repo actions`
- Pick an action to configure (recommend `App Build` followed by `Deploy Firebase`)
- Follow interactive prompts to set action variables and secrets
- Review populated deployment code changes (e.g. create PR via `yarn workflow repo publish`)
- Test actions on merge (or edit action to trigger on current pr branch)

## Follow-up Tasks (to make issue)
Support and template updates will be required for deployments that use encrypted deployments (needs testing)

Actions previously hardcoded firebase config values. As we will want to change these per-deployment they should need to be refactored to be stored in deployment config instead (likely as encrypted setting). It would also be good to make firebase an opt-in feature (handle case where firebase config not defined). As this would be a mid-size refactor best done as part of a separate PR

Once firebase config migrated action templates can be modified to populate from deployment as part of deployment set/build workflow. 

## Git Issues

## Developer Notes
Action templates are stored within the `packages/actions` workspace, and consist of `template.yml` and `config.ts` files.
The `template.yml` stores the main github action template yml, with additional placeholders for variables to populate.
The `config.ts` file 

As it stands actions are only populated once, so if updating templates any deployments using will need to manually reconfigure

Closes #

## Screenshots/Videos
**Example use of CLI to create github action**
![image](https://github.com/IDEMSInternational/parenting-app-ui/assets/10515065/9e94a069-4e79-4884-af84-24358397d23e)

**Example generated action code**
```yaml
##################################################################################
#         About
##################################################################################
# Build and deploy app to firebase
# Supports deploying to custom hosting channel for long-lived preview urls (30d)
# Or using custom hosting target

##################################################################################
#         Configuration (updated per deployment)
##################################################################################

env:
  FIREBASE_PROJECT_ID: "example-firebase-id" # | ID of firebase project used (in case of multiple deployment targets just specify default)
  FIREBASE_HOSTING_CHANNEL: "live" # | Name of channel to deploy to (default 'live' is main site, any other word, e.g. 'pr' will create random temp preview site)
  FIREBASE_HOSTING_TARGET: "" # | Optional override if using multiple hosting target sites (default target project ID)
  FIREBASE_SERVICE_ACCOUNT: ${{secrets.FIREBASE_SERVICE_ACCOUNT}} # | JSON export of firebase service account (from console)
##################################################################################
#         Triggers
# Specify conditions to run workflow. # See more information at:
# https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows
##################################################################################
on:
  push:
    branches:
      - "main"

##################################################################################
#         Main Code
##################################################################################
name: Deploy - Firebase
# Allow only one concurrent deployment, skipping runs queued between the run in-progress and latest queued.
concurrency:
  group: "deploy_firebase"
  cancel-in-progress: true

jobs:
  build_action:
    uses: ./.github/workflows/app-build.yml
    with:
      artifact-name: www

  deploy_dev:
    needs: build_action
    runs-on: ubuntu-latest
    outputs:
      urls: ${{ steps.deploy.outputs.urls }}
    steps:
      # Extract build artifact
      - uses: actions/checkout@v3
      - name: Download Build Artifact
        uses: actions/download-artifact@v3
        with:
          name: www
      - name: Extract Build folder
        run: |
          mkdir www
          tar -xf artifact.tar --directory www

      # Ensure FIREBASE_HOSTING_TARGET set (default fallback to projectId)
      - name: Populate Env
        if: ${{env.FIREBASE_HOSTING_TARGET == ''}}
        run: echo "FIREBASE_HOSTING_TARGET=${{env.FIREBASE_PROJECT_ID}}" >> "$GITHUB_ENV"

      # Create a .firebaserc file mapping any firebase deployment host targets (required if multi-host projects)
      # e.g. {"projects": {"default": "my_app"},"targets": {"my_app": {"hosting": {"my_app_dev":["my_app_dev"]} } }
      - name: Populate Firebase Targets
        run: |
          FIREBASE_RC_TARGETS=$(jq -n \
            --argjson "${{env.FIREBASE_PROJECT_ID}}" \
            '{"hosting":{"${{env.FIREBASE_HOSTING_TARGET}}":["${{env.FIREBASE_HOSTING_TARGET}}"]}}' \
            '$ARGS.named'
            )
          FIREBASE_RC=$(jq -n \
          --argjson projects '{"default":"${{env.FIREBASE_PROJECT_ID}}"}' \
          --argjson targets "$FIREBASE_RC_TARGETS" \
          '$ARGS.named'
          )
          echo $FIREBASE_RC | jq '.'
          echo $FIREBASE_RC > .firebaserc

      # Create a firebase.json file to handle single-page-app hosting redirects
      # e.g. {"hosting": [{"target": "app","public": "www","ignore": ["firebase.json"], "rewrites": [{"source": "**","destination": "/index.html"}] }]}
      - name: Populate Firebase JSON
        run: |
          FIREBASE_JSON_HOSTING=$(jq -n \
          --arg target "${{ env.FIREBASE_HOSTING_TARGET }}" \
          --arg public "www" \
          --argjson ignore '["firebase.json"]' \
          --argjson rewrites '[{"source": "**","destination": "/index.html"}]' \
          '$ARGS.named'
          )
          FIREBASE_JSON=$(jq -n \
          --argjson hosting "[$FIREBASE_JSON_HOSTING]" \
          '$ARGS.named'
          )
          echo $FIREBASE_JSON | jq '.'
          echo $FIREBASE_JSON > firebase.json

      # Deploy to firebase
      - id: deploy
        uses: FirebaseExtended/action-hosting-deploy@v0
        with:
          repoToken: "${{ secrets.GITHUB_TOKEN }}"
          firebaseServiceAccount: "${{ env.FIREBASE_SERVICE_ACCOUNT }}"
          projectId: "${{ env.FIREBASE_PROJECT_ID }}"
          channelId: "${{ env.FIREBASE_HOSTING_CHANNEL }}"
          target: "${{ env.FIREBASE_HOSTING_TARGET }}"
          expires: "30d"
##################################################################################
#         Useful Links
##################################################################################
# https://firebase.google.com/docs/hosting/full-config
# https://firebase.google.com/docs/cli/targets

# https://www.baeldung.com/linux/jq-command-json
# https://jqlang.github.io/jq/manual/
# https://jqplay.org/

# echo $FIREBASE_JSON | jq '.hosting[0]'

```

**Documentation added**
![image](https://github.com/IDEMSInternational/parenting-app-ui/assets/10515065/cc0dee79-4334-4b6d-9b40-59544d726fe9)

jfmcquade commented on packages/actions/templates/app-build/template.yml:
```diff
@@ -0,0 +1,77 @@
+
+##################################################################################
+#         Main Code
+##################################################################################
```
Would be good to add a `name`, as with the "Deploy - Firebase" action, so that it gets named in Github Actions 
<img width="328" alt="Screenshot 2023-06-27 at 10 37 56" src="https://github.com/IDEMSInternational/parenting-app-ui/assets/64838927/0e19fa09-14e5-4e8d-bd61-16c5af89bab1">

jfmcquade commented:
Very nice to have. 

I've tested the `App Build` and `Deploy Firebase` actions. Both configuring them for my deployment (repo [here](https://github.com/jfmcquade/app-debug)), as well as the [actions themselves](https://github.com/jfmcquade/app-debug/actions/runs/5388371087/jobs/9781165799) seem to work as expected.
> The `build` workflow run initially failed with an error because `src/assets/app_data/sheets/contents.json` was missing the `generator` property, I suppose because the contents file was generated using a legacy parser. This file wasn't updated when I ran `yarn workflow sync`, but I fixed the issue by manually copying over the contents of `.idems_app/deployments/jfmcquade-app-debug/app_data/sheets/contents.json` into  `src/assets/app_data/sheets/contents.json`.

There's some assumed knowledge of Firebase, but I think that's fine as we wouldn't expect inexperienced authors to be setting up these actions. As you note in the dev notes, if the core templates for the actions are changed then those changes will need to be manually reflected in all content repos that use the actions. Again, the fact that this would need to be done by a dev doesn't seem like an issue, and we wouldn't expect to make such changes regularly. But it is a motivation for getting the core templates to be as advanced as possible before copying them over to other repos.

esmeetewinkel reviewed (commented):
I tried using this to set up the app-build and deploy-firebase workflows on the app-debug-content repo (after first [removing the workflows that were there](https://github.com/IDEMSInternational/app-debug-content/pull/19)). 

**App build**
<img width="376" alt="image" src="https://github.com/IDEMSInternational/parenting-app-ui/assets/74557272/9d0a9a01-c435-493b-9fdf-25e238cf7b8e">

**Deploy to Firebase**
<img width="376" alt="image" src="https://github.com/IDEMSInternational/parenting-app-ui/assets/74557272/30424a93-39e1-4da1-8039-2ccd034a03e1">

I got stuck when trying to push my changes to GitHub using `yarn workflow repo publish`, getting the error below.
<img width="861" alt="image" src="https://github.com/IDEMSInternational/parenting-app-ui/assets/74557272/46ffa9ab-836c-46f0-b500-d657c3fac833">
Not sure if this is my lack of understanding or an actual bug!

chrismclarke commented:
@jfmcquade 
I've added the build action name, and also added a script that populates some custom metadata in a tag at the top of each generated template so that in the future we could perhaps look to automate the process of updating deployment repo actions

![image](https://github.com/IDEMSInternational/parenting-app-ui/assets/10515065/0c59fa30-092f-4746-862e-35f19b37a4ae)

Agreed that we should expect a certain level of technical knowledge for anyone using custom automation actions, could do more in the future like improving docs, publishing secrets etc. but for now feels like already going pretty far to facilitate what is otherwise a very complex/technical challenge

chrismclarke commented:
@esmeetewinkel 

> I got stuck when trying to push my changes to GitHub using yarn workflow repo publish, getting the error below.
> image
> Not sure if this is my lack of understanding or an actual bug!

Taking a quick google it seems like an issue related more to git/github than our specific code. I can see one potential solution here: https://jessehouwing.net/cannot-push-to-github-without-workflow-scope/

Otherwise you might need to google search the issue and see if there are any others, not currently possible for me to recreate on my side unfortunately

jfmcquade reviewed (approved):
I think this looks good and should be ready to merge. We want authors to be able to update the web preview of the facilitator app soon, so this will be useful

esmeetewinkel reviewed (commented):
Thanks @chrismclarke, I was able to resolve the OAuth issue and add the two actions to the app-debug-content repo.

The build action is [failing](https://github.com/IDEMSInternational/app-debug-content/actions/runs/5456001878) with the error message Cannot find module './firebaseConfig' or its corresponding type declarations. - any insights on that?
image

jfmcquade commented:
> The build action is [failing](https://github.com/IDEMSInternational/app-debug-content/actions/runs/5456001878) with the error message `Cannot find module './firebaseConfig' or its corresponding type declarations.` - any insights on that?

That's because we also need to add the the `FIREBASE_CONFIG` secret to the repo [here](https://github.com/IDEMSInternational/app-debug-content/settings/secrets/actions). This secret should contain the equivalent contents of the file that we have locally at `src/environments/firebaseConfig.ts`. I've added it now, so hopefully the [build and deployment](https://github.com/IDEMSInternational/app-debug-content/actions/runs/5456001878) should work.

The instructions say "Action requires Github Secrets to be manually configured", but perhaps they could be clearer about what that means? Either way, as long as we know how to do it that should be sufficient

chrismclarke commented:
> The instructions say "Action requires Github Secrets to be manually configured", but perhaps they could be clearer about what that means? Either way, as long as we know how to do it that should be sufficient

Yeah at this point I'm reluctant to go into too much additional detail on exactly how to manage each action as somewhat niche/subject-to-change, right now the message shows as follows

![image](https://github.com/IDEMSInternational/parenting-app-ui/assets/10515065/cf069060-35b6-4f24-b772-6cdae90d9467)

Can I suggest we add to the list of future documentation any additional steps required to support authors configuring these variables? 

As for the case of firebase config, I think we probably want to replace the use of github secrets with deployment variable instead (like we do for supabase config). I'll add a low-priority follow-up issue #1976
